proposals/p2138.md

```diff
@@ -0,0 +1,59 @@
+# Checked and template generic terminology
+
+<!--
+Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+Exceptions. See /LICENSE for license information.
+SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+-->
+
+[Pull request](https://github.com/carbon-language/carbon-lang/pull/2138)
+
+<!-- toc -->
+
+## Table of contents
+
+-   [Abstract](#abstract)
+-   [Problem](#problem)
+-   [Proposal](#proposal)
+-   [Rationale](#rationale)
+
+<!-- tocstop -->
+
+## Abstract
+
+Change terminology from "generics" and "templates" to "checked generics" and
+"template generics". Afterwards, "generics" will be an umbrella term that
+include templates.
+
+## Problem
+
+The C++ community does "generic programming" using templates in C++, and so
+thinks of templates as a mechanism for implementing generics. In Carbon,
+templates and other generic features are more similar than they are different,
+so it is worthwhile to have an easy way to talk about things that apply to both.
+
+## Proposal
+
+Change terminology from "generics" and "templates" to "checked generics" and
+"template generics". Afterwards, "generics" will be an umbrella term that
+include templates.
```

Comment:
Given that we don't require a generic to have only checked parameters or only unchecked parameters, I think that in cases where we want to be precise, we won't want to talk about "checked generics" and "template generics" but instead will want to talk about individual parameters. So I think we're also going to need terminology for the different kinds of parameters, and it's not clear to me what terms we would use based on this new terminology.

I don't especially like any of the options that spring to mind:

-   "checked parameters" and "template parameters": "checked parameters" isn't very clear that we're talking about a generic parameter rather than a runtime parameter
-   "checked parameters" and "unchecked parameters": these also sound like runtime parameters
-   "checked generic parameters" and "template generic parameters": too long
-   "checked generic parameters" and "unchecked generic parameters": also too long
-   "generic parameters" and "template parameters": works prior to this proposal, but doesn't properly address the fact that a template is a kind of generic

I'm also a little concerned that if we use the terms "checked generic parameter" and "template generic parameter", people are going to shorten the former to "generic parameter" and the latter to "template parameter", losing the unifying term that covers both.

Reply:
> "checked parameters" and "template parameters": "checked parameters" isn't very clear that we're talking about a generic parameter rather than a runtime parameter

Are checked generic parameters always going to be interface-based? Instead of "checked", could we say "interface parameters" and "template parameters"?

Reply:
Personally, I'm down with "checked parameters" and "template parameters".

I map "checked parameters" to "definition checked parameters" or "definition type checked parameters", and then shorten it, and this seems to nicely contrast with "late bound template parameters"

🤷🏻 Not sure there is perfect terminology here, but I think "checked ..." and "template ..." works well enough.

Reply:
> Are checked generic parameters always going to be interface-based? Instead of "checked", could we say "interface parameters" and "template parameters"?

Generic parameters will commonly be interface-based, but not always. Template parameters may also have interface constraints.

Reply:
I think it's important that we don't omit the word "generic" except when it's clear from context. I find "checked parameter" to be too unclear to use as a context-free term.

I think the solution that I find I like the most here is:

-   As per this proposal, we have "checked generics" and "template generics", collectively "generics", but, as noted at the start of this thread, the first two terms are approximate given that a generic might have both kinds of parameter.
-   We have "checked generic parameters" and "template generic parameters", collectively "generic parameters".
-   In contexts where it's already clear that we're talking about a generic parameter, it's OK to say simply "checked parameter" or to use "checked" as an adjective: "A generic type parameter has a constraint that describes an interface that the type implements. For a checked parameter, only the names in that interface are available for member access into the type; for a template parameter, member names of the type are available too."
-   Except in places where we want to be especially clear that a template parameter is a kind of generic parameter, such as when introducing those notions, we assume that it's clear that a template parameter is a kind of generic parameter, so it's generally OK to say simply "template parameter" or "template" instead of "template generic parameter" or "template generic".
-   It's preferable to use the collective terms "generic" or "generic parameter" when the kind is not important.

Does that seem reasonable? I think that's consistent with the direction of this proposal.

```suggestion
Change terminology from "generics" and "templates" to "checked generics" and
"template generics". Afterwards, "generics" will be an umbrella term that
include templates.

Similarly, for parameters, use "checked generic parameter" and "template generic
parameter", with the umbrella term "generic parameter" describing both kinds. The
word "generic" can be omitted from these terms if it's clear from context, which
will usually be the case for template parameters.
```